File: python/covertable/pict/weights.py
"""Helpers for converting human-friendly weight specifications."""


def weights_by_value(factors, value_weights):
    """Convert value-keyed weights to index-keyed weights.

    Useful when you want to specify weights by value rather than by index
    position.

    Example::

        weights_by_value(
            {"Browser": ["Chrome", "Firefox", "Safari"]},
            {"Browser": {"Chrome": 10, "Safari": 5}},
        )
        # → {"Browser": {0: 10, 2: 5}}
    """
    result = {}
    for key, vw in value_weights.items():
        values = factors.get(key)
        if values is None:
            continue
        index_weights = {}
        for value, weight in vw.items():
            for i, v in enumerate(values):
                if str(v) == str(value):
                    index_weights[i] = weight
                    break
        if index_weights:
            result[key] = index_weights
    return result

File: typescript/src/pict/weights.ts
import type { PictFactorsType } from './types';

/**
 * Convert value-keyed weights to index-keyed weights for use with `OptionsType.weights`.
 * Useful when you want to specify weights by value rather than by index position.
 *
 * @example
 *   weightsByValue(
 *     { Browser: ["Chrome", "Firefox", "Safari"] },
 *     { Browser: { Chrome: 10, Safari: 5 } },
 *   )
 *   // → { Browser: { 0: 10, 2: 5 } }
 */
export function weightsByValue<T extends PictFactorsType>(
  factors: T,
  valueWeights: { [factorKey: string]: { [value: string]: number } },
): { [factorKey: string]: { [index: number]: number } } {
  const result: { [factorKey: string]: { [index: number]: number } } = {};
  for (const [key, vw] of Object.entries(valueWeights)) {
    const values = factors[key];
    if (!values) continue;
    const indexWeights: { [index: number]: number } = {};
    for (const [valueStr, weight] of Object.entries(vw)) {
      const idx = values.findIndex(v => String(v) === valueStr);
      if (idx >= 0) indexWeights[idx] = weight;
    }
    if (Object.keys(indexWeights).length > 0) {
      result[key] = indexWeights;
    }
  }
  return result;
}
